parser/README.md

```diff
@@ -0,0 +1,134 @@
+
+An alternative is to re-implement SQLFlow for each SQL engine as their extensions or UDF (user-defined function). This approach is equally intractable as there are so many engines to support.
+
+## Overall Design
+
+We design SQLFlow as a wrapper of these SQL engines, which means that SQL statements come directly to SQLFlow and SQLFlow decides if it follows the dialect syntax of the specified SQL engine. If so, SQLFlow proxies the statement to the SQL engine, otherwise, SQLFlow translates it into a Python program, a.k.a., the submitter program, which calls the SQL engine and the AI engine to train or to predict.
+
+With this design, the key challenge becomes to judge if a SQL statement is acceptable by a SQL engine's parser. We propose the following solution: 
+
+```go
+func Parse(sql string) (acceptable bool, err error) {
+    // Where SQL is acceptable by the original engine, no error occurred.
+    err := sql_engine.Parse(sql)
+    if err == nil {
+        return true, ni
+    }
+    
+    // Error message from a parser should contain error position.
+    pos := parseErrorPosition(err) 
+    leftPart = sql[:pos]
+    rightPart = sql[pos:]
+
+    errLeft := sql_engine.Parse(leftPart)
+    // In this case, the SQL is not acceptable due to the syntax error
+    if err != nil {
+       return false, err 
+    }
+
+    // If leftPart is acceptable, it is a legitimate  SELECT statement. We then try right part with SQLFlow parser. 
+    errRight := SQLFlow.Parse(rightPart)
+    if err != nil {
+        return false, errRight 
+    }
+
+    // The left part is a SELECT and the right part is TRAIN or PREDICT.
+    return false, nil 
+}
+```
+
+This procedure has two assumptions: (1) a SQL engine has a parser, denoted by sql_engine.Parse, and (2) the parser reports the position of syntax errors if there is any.  Both assumptions are reasonable. However, a key question here is how to call parsers of various SQL engines. For open source SQL engines, this doesn't seem an issue. Here we list parsers used by some well-known engines:
+
+| SQL Engine | Parser  |
+|------------|---------|
+| Beam       | Calcite |
+| Flink      | Calcite |
+| Storm      | Calcite |
+| Hive       | Hplsql.g4 |
+| MySQL      | sql_parse.{h,cc} |
+
+It is notable that MySQL doesn't provide sufficient documentation on how to call its parser, so we call a Go implementation of MySQL's SQL parser by PingCap https://github.com/pingcap/parser instead. For proprietary engines, we have a plan to ask the licensing of their parsers.
+
+
+## The External Parser Abstraction
+
+The above example program calls `sql_engine.Parse` twice before calling `SQLFlow.Parse`.  In practice, because Calcite parser is in Java, SQLFlow is a Go program, to enable SQLFlow calling Calcite parser, we have to wrap Calcite parser up into a gRPC server.  It is time-consuming to make an RPC call, so we pack the two calls to `sql_engine.Parse` into one:
```

Comment:
Agree with `packing the two calls into one` to reduce time-consuming.
But I would suggest keeping the parses as simple as possible:
1. for a correct extended SQL: latency is mainly caused by the execution phase
2. I'm not sure if `SQLFow` would expand keywords like `yet_another_train`. If yes: we should update `SQLFlow` & `external parsers`, which means more complicate for deploy.